### DISCUSSION TITLE
What's the main difference between TinyMPC and ReLU-QP?

### DISCUSSION THREAD

**Discussion by FenglongSong** [TinyMPC/TinyMPC#35] [Q&A]
Hi all,

Thanks for your great work. It's amazing. I'm wondering what are the main differences between TinyMPC and ReLU-QP? As what I can see, both works rely on ADMM and precompute matrix inversions to avoid solving linear systems online. Would appreciate your answers! Thanks a lot!

**Comment by xkhainguyen** [comment]
Hi,

Thank you for the interesting question. Both projects share the same core optimization algorithm—ADMM with matrix inverse precomputation. However, we have different perspectives when it comes to implementations and use cases: parallelizable on GPUs and compact/miniature on MCUs.

Specifically, TinyMPC (natively in C/C++) uses small, dense matrices, allowing us to operate efficiently on very compute-limited MCUs (in terms of frequency and memory). Additionally, we support high-level interfaces for low-level code generation and advanced features such as second-order cone constraints on both input and state. TinyMPC can totally be used to solve problems featured in the ReLU-QP paper.

**Reply by FenglongSong** [reply]
Nice, thanks for your quick answer! It's very clear!
